google/cloud/spanner/internal/transaction_impl_test.cc

```diff
@@ -110,20 +111,20 @@ class Client {
 // switch the selector to use the allocated transaction ID.  Here we use
 // the pre-assigned transaction ID after checking the read timestamp.
 ResultSet Client::Read(SessionHolder& session,
-                       absl::optional<TransactionSelector>& selector,
+                       StatusOr<TransactionSelector>& selector,
                        std::int64_t seqno, std::string const&, KeySet const&,
                        std::vector<std::string> const&) {
   bool fail_with_throw = false;
   if (!selector) {
     std::unique_lock<std::mutex> lock(mu_);
-    switch (mode_) {
-      case Mode::kReadSucceeds:  // visits never valid
-        break;
-      case Mode::kReadFailsAndTxnRemainsBegin:  // visits always valid
-      case Mode::kReadFailsAndTxnInvalidated:
-        ++valid_visits_;
-        fail_with_throw = (valid_visits_ % 2 == 0);
-        break;
+    // we should only get here in "transaction invalidated" mode.
+    if (mode_ == Mode::kReadFailsAndTxnInvalidated) {
+      EXPECT_EQ(selector.status(), failed_txn_status_);
+      // visits always valid
+      ++valid_visits_;
+      fail_with_throw = (valid_visits_ % 2 == 0);
+    } else {
+      ADD_FAILURE() << "Unexpected mode_ " << static_cast<int>(mode_);
```

Comment:
This would have been (and still will be) picked up by the expectation on `valid_visits_`.  So, unless we want to add failures for all the other places we don't increment `valid_visits_` during this visitor, I would suggest not adding this one.